File: wagerproof-ios-native/Wagerproof/Features/Chat/WagerBotActionPreview.swift
import SwiftUI
import WagerproofDesign
import WagerproofModels

/// Renders a `present_analysis` widget block as a rich inline card.
/// Wagerproof's analogue of Honeydew's `AssistantV2ActionPreview` — same
/// "full-width tile inside the assistant bubble" layout, but the
/// content is sport-specific analysis (matchup overview, model
/// projections, polymarket odds, public betting splits, injuries,
/// betting trends, weather) instead of recipe / meal-plan / grocery
/// previews.
///
/// One widget per `WagerBotChatWidget` block; multiple widgets render
/// stacked. Tapping anywhere on the tile fires `onTap` which opens the
/// sport game sheet for the parent game.
struct WagerBotActionPreview: View {
    let widget: WagerBotChatWidget
    let ui: WagerBotUiTokens
    var onTap: (() -> Void)?

    var body: some View {
        Button {
            onTap?()
        } label: {
            VStack(alignment: .leading, spacing: 10) {
                header
                if let analysis = widget.analysis, !analysis.isEmpty {
                    WagerBotMarkdownText(
                        analysis,
                        baseFont: .system(size: 13, weight: .medium),
                        primaryColor: ui.primaryText,
                        secondaryColor: ui.mutedText
                    )
                }
                widgetBody
                footer
            }
            .padding(14)
            .frame(maxWidth: .infinity, alignment: .leading)
            .background(ui.hintChipBackground)
            .clipShape(RoundedRectangle(cornerRadius: 14, style: .continuous))
            .overlay(
                RoundedRectangle(cornerRadius: 14, style: .continuous)
                    .stroke(ui.borderColor, lineWidth: 1)
            )
        }
        .buttonStyle(.plain)
    }

    // MARK: - Header

    private var header: some View {
        HStack(spacing: 8) {
            Image(systemName: headerIcon)
                .font(.system(size: 12, weight: .bold))
                .foregroundStyle(ui.accent)
            Text(headerLabel)
                .font(.system(size: 12, weight: .semibold))
                .foregroundStyle(ui.accent)
            Spacer(minLength: 0)
            Text(widget.sport.uppercased())
                .font(.system(size: 10, weight: .bold))
                .foregroundStyle(ui.mutedText)
        }
    }

    private var headerLabel: String {
        if let title = widget.title, !title.isEmpty { return title }
        switch widget.widgetType {
        case "matchup":          return "Matchup overview"
        case "model_projection": return "Model projection"
        case "polymarket":       return "Polymarket"
        case "public_betting":   return "Public betting"
        case "injuries":         return "Injuries"
        case "betting_trends":   return "Betting trends"
        case "weather":          return "Weather"
        default:                 return widget.widgetType
            .replacingOccurrences(of: "_", with: " ")
            .capitalized
        }
    }

    private var headerIcon: String {
        switch widget.widgetType {
        case "matchup":          return "rectangle.split.2x1.fill"
        case "model_projection": return "wand.and.stars"
        case "polymarket":       return "chart.line.uptrend.xyaxis"
        case "public_betting":   return "person.2.fill"
        case "injuries":         return "cross.fill"
        case "betting_trends":   return "chart.bar.fill"
        case "weather":          return "cloud.sun.fill"
        default:                 return "chart.bar.doc.horizontal.fill"
        }
    }

    // MARK: - Body (widget-specific KV grid)

    /// Decode the widget's `data` envelope and render the most relevant
    /// rows for the widget kind. Falls back to a generic top-level k/v
    /// scan when the type is unknown so new widget kinds at least render
    /// something useful while the iOS side catches up.
    @ViewBuilder
    private var widgetBody: some View {
        if let data = widget.dataJSON,
           let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any] {
            switch widget.widgetType {
            case "model_projection":
                modelProjectionRows(obj)
            case "polymarket":
                polymarketRows(obj)
            case "public_betting":
                publicBettingRows(obj)
            case "injuries":
                injuriesRows(obj)
            case "betting_trends":
                trendsRows(obj)
            case "weather":
                weatherRows(obj)
            case "matchup":
                matchupRows(obj)
            default:
                genericRows(obj)
            }
        }
    }

    // MARK: - Rendering helpers (per widget kind)

    @ViewBuilder
    private func modelProjectionRows(_ obj: [String: Any]) -> some View {
        VStack(alignment: .leading, spacing: 6) {
            if let pick = obj["spread_pick"] as? String {
                kvRow("Spread pick", pick)
            }
            if let prob = obj["ml_prob"] as? Double {
                kvRow("Moneyline", String(format: "%.1f%%", prob * 100))
            }
            if let ouPick = obj["ou_pick"] as? String {
                let ouLine = obj["over_under"] as? Double
                kvRow("Total", ouLine.map { "\(ouPick) \(String(format: "%.1f", $0))" } ?? ouPick.capitalized)
            }
            if let edge = obj["spread_edge"] as? Double {
                kvRow("Edge", String(format: "%+.1f pts", edge))
            }
        }
    }

    @ViewBuilder
    private func polymarketRows(_ obj: [String: Any]) -> some View {
        VStack(alignment: .leading, spacing: 6) {
            if let home = obj["home_implied_prob"] as? Double {
                kvRow("Home implied", String(format: "%.1f%%", home * 100))
            }
            if let away = obj["away_implied_prob"] as? Double {
                kvRow("Away implied", String(format: "%.1f%%", away * 100))
            }
            if let volume = obj["volume_usd"] as? Double {
                kvRow("Volume", String(format: "$%.0fK", volume / 1000))
            }
        }
    }

    @ViewBuilder
    private func publicBettingRows(_ obj: [String: Any]) -> some View {
        VStack(alignment: .leading, spacing: 6) {
            if let pct = obj["home_bet_pct"] as? Double {
                kvRow("Home bet %", String(format: "%.0f%%", pct))
            }
            if let pct = obj["away_bet_pct"] as? Double {
                kvRow("Away bet %", String(format: "%.0f%%", pct))
            }
            if let pct = obj["home_money_pct"] as? Double {
                kvRow("Home $", String(format: "%.0f%%", pct))
            }
        }
    }

    @ViewBuilder
    private func injuriesRows(_ obj: [String: Any]) -> some View {
        VStack(alignment: .leading, spacing: 6) {
            if let list = obj["players"] as? [[String: Any]] {
                ForEach(Array(list.prefix(4).enumerated()), id: \.offset) { _, p in
                    let name = (p["name"] as? String) ?? ""
                    let status = (p["status"] as? String) ?? ""
                    kvRow(name, status)
                }
            }
        }
    }

    @ViewBuilder
    private func trendsRows(_ obj: [String: Any]) -> some View {
        VStack(alignment: .leading, spacing: 6) {
            ForEach(Array(obj.prefix(5)), id: \.key) { (key, value) in
                kvRow(humanize(key), stringify(value))
            }
        }
    }

    @ViewBuilder
    private func weatherRows(_ obj: [String: Any]) -> some View {
        VStack(alignment: .leading, spacing: 6) {
            if let temp = obj["temperature"] as? Double {
                kvRow("Temperature", "\(Int(temp))°F")
            }
            if let wind = obj["wind_speed"] as? Double {
                kvRow("Wind", "\(Int(wind)) mph")
            }
            if let cond = obj["conditions"] as? String {
                kvRow("Conditions", cond)
            }
        }
    }

    @ViewBuilder
    private func matchupRows(_ obj: [String: Any]) -> some View {
        VStack(alignment: .leading, spacing: 6) {
            if let away = obj["away_team"] as? String,
               let home = obj["home_team"] as? String {
                kvRow("Matchup", "\(away) @ \(home)")
            }
            if let time = obj["game_time"] as? String {
                kvRow("Time", time)
            }
            if let venue = obj["venue"] as? String {
                kvRow("Venue", venue)
            }
        }
    }

    @ViewBuilder
    private func genericRows(_ obj: [String: Any]) -> some View {
        VStack(alignment: .leading, spacing: 6) {
            ForEach(Array(obj.prefix(5)), id: \.key) { (key, value) in
                kvRow(humanize(key), stringify(value))
            }
        }
    }

    private func kvRow(_ label: String, _ value: String) -> some View {
        HStack(alignment: .firstTextBaseline) {
            Text(label)
                .font(.system(size: 12, weight: .medium))
                .foregroundStyle(ui.mutedText)
            Spacer(minLength: 8)
            Text(value)
                .font(.system(size: 13, weight: .semibold))
                .foregroundStyle(ui.primaryText)
                .multilineTextAlignment(.trailing)
                .lineLimit(2)
        }
    }

    private var footer: some View {
        HStack(spacing: 6) {
            Spacer(minLength: 0)
            Text("View game details")
                .font(.system(size: 11, weight: .semibold))
                .foregroundStyle(ui.accent)
            Image(systemName: "arrow.up.right")
                .font(.system(size: 10, weight: .bold))
                .foregroundStyle(ui.accent)
        }
    }

    // MARK: - Helpers

    private func humanize(_ key: String) -> String {
        key.replacingOccurrences(of: "_", with: " ").capitalized
    }

    private func stringify(_ value: Any) -> String {
        if let s = value as? String { return s }
        if let n = value as? NSNumber { return n.stringValue }
        if let b = value as? Bool { return b ? "Yes" : "No" }
        if let arr = value as? [Any] { return "\(arr.count) items" }
        if let _ = value as? [String: Any] { return "..." }
        return String(describing: value)
    }
}

File: wagerproof-android-native/app/src/main/java/com/wagerproof/app/features/chat/WagerBotActionPreview.kt
package com.wagerproof.app.features.chat

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.wagerproof.core.models.WagerBotChatWidget
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.booleanOrNull
import kotlinx.serialization.json.doubleOrNull
import java.time.Instant
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.util.Locale

/**
 * Port of iOS `WagerBotActionPreview.swift`. Renders a `present_analysis`
 * widget (legacy V1 path) as a rich inline card: header + optional markdown +
 * a type-dispatched KV grid + a "View game details" footer. Tap opens the game
 * sheet via [onTap].
 */
@Composable
fun WagerBotActionPreview(
    widget: WagerBotChatWidget,
    ui: WagerBotUiTokens,
    modifier: Modifier = Modifier,
    onTap: (() -> Unit)? = null,
) {
    val shape = RoundedCornerShape(14.dp)
    Column(
        modifier = modifier
            .fillMaxWidth()
            .clip(shape)
            .background(ui.hintChipBackground)
            .border(1.dp, ui.borderColor, shape)
            .then(if (onTap != null) Modifier.clickable { onTap() } else Modifier)
            .padding(14.dp),
        verticalArrangement = Arrangement.spacedBy(10.dp),
    ) {
        Header(widget, ui)
        widget.analysis?.takeIf { it.isNotEmpty() }?.let {
            WagerBotMarkdownText(
                text = it,
                baseFontSize = 13.sp,
                baseFontWeight = FontWeight.Medium,
                primaryColor = ui.primaryText,
                secondaryColor = ui.mutedText,
            )
        }
        WidgetBody(widget, ui)
        Footer(ui)
    }
}

@Composable
private fun Header(widget: WagerBotChatWidget, ui: WagerBotUiTokens) {
    Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(8.dp)) {
        Icon(chatIcon(headerIcon(widget.widgetType)), contentDescription = null, tint = ui.accent, modifier = Modifier.padding(0.dp).then(Modifier).let { it }.let { Modifier }.let { Modifier })
        Text(headerLabel(widget), fontSize = 12.sp, fontWeight = FontWeight.SemiBold, color = ui.accent)
        Spacer(Modifier.weight(1f))
        Text(widget.sport.uppercase(), fontSize = 10.sp, fontWeight = FontWeight.Bold, color = ui.mutedText)
    }
}

private fun headerLabel(widget: WagerBotChatWidget): String {
    widget.title?.takeIf { it.isNotEmpty() }?.let { return it }
    return when (widget.widgetType) {
        "matchup" -> "Matchup overview"
        "model_projection" -> "Model projection"
        "polymarket" -> "Polymarket"
        "public_betting" -> "Public betting"
        "injuries" -> "Injuries"
        "betting_trends" -> "Betting trends"
        "weather" -> "Weather"
        else -> widget.widgetType.replace('_', ' ').replaceFirstChar { it.titlecase(Locale.US) }
    }
}

private fun headerIcon(type: String): String = when (type) {
    "matchup" -> "rectangle.split.2x1.fill"
    "model_projection" -> "wand.and.stars"
    "polymarket" -> "chart.line.uptrend.xyaxis"
    "public_betting" -> "person.2.fill"
    "injuries" -> "cross.fill"
    "betting_trends" -> "chart.bar.fill"
    "weather" -> "cloud.sun.fill"
    else -> "chart.bar.doc.horizontal.fill"
}

@Composable
private fun WidgetBody(widget: WagerBotChatWidget, ui: WagerBotUiTokens) {
    val obj = widget.dataJson as? JsonObject ?: return
    val rows: List<Pair<String, String>> = when (widget.widgetType) {
        "model_projection" -> modelProjectionRows(obj)
        "polymarket" -> polymarketRows(obj)
        "public_betting" -> publicBettingRows(obj)
        "injuries" -> injuriesRows(obj)
        "betting_trends" -> genericRows(obj)
        "weather" -> weatherRows(obj)
        "matchup" -> matchupRows(obj)
        else -> genericRows(obj)
    }
    if (rows.isEmpty()) return
    Column(verticalArrangement = Arrangement.spacedBy(6.dp)) {
        rows.forEach { (label, value) -> KvRow(label, value, ui) }
    }
}

@Composable
private fun KvRow(label: String, value: String, ui: WagerBotUiTokens) {
    Row(verticalAlignment = Alignment.Top) {
        Text(label, fontSize = 12.sp, fontWeight = FontWeight.Medium, color = ui.mutedText)
        Spacer(Modifier.weight(1f).widthIn(min = 8.dp))
        Text(
            text = value,
            fontSize = 13.sp,
            fontWeight = FontWeight.SemiBold,
            color = ui.primaryText,
            textAlign = TextAlign.End,
            maxLines = 2,
        )
    }
}

@Composable
private fun Footer(ui: WagerBotUiTokens) {
    Row(Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(6.dp)) {
        Spacer(Modifier.weight(1f))
        Text("View game details", fontSize = 11.sp, fontWeight = FontWeight.SemiBold, color = ui.accent)
        Icon(chatIcon("arrow.up.right"), contentDescription = null, tint = ui.accent, modifier = Modifier.padding(0.dp))
    }
}

// MARK: - Per-widget row builders (mirror the iOS decoders)

private fun JsonObject.dbl(key: String): Double? =
    ((this[key] as? JsonPrimitive))?.let { if (it.isString) it.content.toDoubleOrNull() else it.doubleOrNull }
private fun JsonObject.str(key: String): String? =
    (this[key] as? JsonPrimitive)?.takeIf { it.isString }?.content

private fun modelProjectionRows(obj: JsonObject): List<Pair<String, String>> = buildList {
    obj.str("spread_pick")?.let { add("Spread pick" to it) }
    obj.dbl("ml_prob")?.let { prob ->
        // 0-1 fraction for some sports, already-scaled percent for others (MLB).
        add("Moneyline" to String.format("%.1f%%", if (prob <= 1) prob * 100 else prob))
    }
    obj.str("ou_pick")?.let { ouPick ->
        val line = obj.dbl("over_under")
        add("Total" to (line?.let { "$ouPick ${String.format("%.1f", it)}" } ?: ouPick.replaceFirstChar { c -> c.titlecase(Locale.US) }))
    }
    obj.dbl("spread_edge")?.let { add("Edge" to String.format("%+.1f pts", it)) }
}

private fun polymarketRows(obj: JsonObject): List<Pair<String, String>> = buildList {
    obj.dbl("home_implied_prob")?.let { add("Home implied" to String.format("%.1f%%", it * 100)) }
    obj.dbl("away_implied_prob")?.let { add("Away implied" to String.format("%.1f%%", it * 100)) }
    obj.dbl("volume_usd")?.let { add("Volume" to String.format("$%.0fK", it / 1000)) }
}

private fun publicBettingRows(obj: JsonObject): List<Pair<String, String>> = buildList {
    obj.dbl("home_bet_pct")?.let { add("Home bet %" to String.format("%.0f%%", it)) }
    obj.dbl("away_bet_pct")?.let { add("Away bet %" to String.format("%.0f%%", it)) }
    obj.dbl("home_money_pct")?.let { add("Home $" to String.format("%.0f%%", it)) }
}

private fun injuriesRows(obj: JsonObject): List<Pair<String, String>> {
    val list = obj["players"] as? JsonArray ?: return emptyList()
    return list.take(4).mapNotNull { el ->
        val p = el as? JsonObject ?: return@mapNotNull null
        (p.str("name") ?: "") to (p.str("status") ?: "")
    }
}

private fun weatherRows(obj: JsonObject): List<Pair<String, String>> = buildList {
    obj.dbl("temperature")?.let { add("Temperature" to "${it.toInt()}°F") }
    obj.dbl("wind_speed")?.let { add("Wind" to "${it.toInt()} mph") }
    obj.str("conditions")?.let { add("Conditions" to it) }
}

private fun matchupRows(obj: JsonObject): List<Pair<String, String>> = buildList {
    val away = obj.str("away_team")
    val home = obj.str("home_team")
    if (away != null && home != null) add("Matchup" to "$away @ $home")
    obj.str("game_time")?.let { add("Time" to formatGameTime(it)) }
    obj.str("venue")?.let { add("Venue" to it) }
}

private fun genericRows(obj: JsonObject): List<Pair<String, String>> =
    obj.entries.take(5).map { (k, v) -> humanize(k) to stringify(v) }

private fun humanize(key: String): String =
    key.replace('_', ' ').replaceFirstChar { it.titlecase(Locale.US) }

private fun stringify(value: JsonElement): String = when (value) {
    is JsonPrimitive -> {
        if (value.isString) value.content
        else value.booleanOrNull?.let { if (it) "Yes" else "No" } ?: value.content
    }
    is JsonArray -> "${value.size} items"
    is JsonObject -> "..."
    else -> value.toString()
}

/** ISO-8601 → "EEE, MMM d · h:mm a". Non-ISO strings pass through. */
private fun formatGameTime(raw: String): String {
    val instant = runCatching { Instant.parse(raw) }.getOrNull() ?: return raw
    val fmt = DateTimeFormatter.ofPattern("EEE, MMM d · h:mm a", Locale.US).withZone(ZoneId.systemDefault())
    return fmt.format(instant)
}
